### PR TITLE
Dev #16784: Revert vuejs top bars

### PR DESCRIPTION
**Pull request by gabrieljenik**
Integrated the new top bar system into the new controllers.
Needed a way for sharing topBars among views and also a general framework where to place common topBar parts (ease maintenance).

Key parts:
LayoutHelper:renderTopbar() called from layouts.
Base top bar with Left and Right side configurable.
Totally overridable if needed.
Implemented it for email templates.

WIP
Next step survey toolbar, which is shared among many screens.

### CODE REVIEW

**Comment by Trischi80** [comment]
Hey Garbiel,

the idea of having a basetopbar template is really nice. And also the seperation of left and right buttons! I like it and will bring it up to discussion. I'm not sure if it was planed/scheduled to change all topbars right now...So maybe we have to stay a little bit longer with the vuejs part and the new topbar  structure...

Greetings Patricia

**Comment by gabrieljenik** [comment]
> I'm not sure if it was planed/scheduled to change all topbars right now

Hi Patricia. Yes, I believe that's the plan.

@olleharstedt Can you shed us some light here?

**Comment by olleharstedt** [comment]
Well, as you know I want an OOP approach, but any intermediate improvement is welcome. :)

**Comment by olleharstedt** [comment]
> I'm not sure if it was planed/scheduled to change all topbars right now

Dunno exactly, but why not, when we've started already?

**Comment by gabrieljenik** [comment]
We have started here: https://bugs.limesurvey.org/view.php?id=16702 and then here https://bugs.limesurvey.org/view.php?id=16784 

The top bars were supposed to be replaced, so we didn't put a lot of effort on adapting/reviewing the topbar for ported screens. Then, we started replacing them, working on some ported screens. While doing that we were building a general topbar replacement.

Did I get something wrong?

**Comment by gabrieljenik** [comment]
Pushed WIP from the surveyTopbar.

# Some details about topBar

topBar is an array with (so far) the following parameters:

- name: Name of the view used for rendering the topbar.

  Some already created views:
  + baseTopbar_view.
   This is a blank topBar which allows left and right subviews.

  + surveyTopbar_view.
   This is a topBar which already includes base buttons for the survey being used.
   Base buttons will be linked to the survey

- leftSideView and rightSideView: views used for rendering the left side and right side of the topbar.

- sid: When using surveyTopbar_view, this is surveyId of the survey being used.
 If not specifically set, the helper will look for the survey id from the following:
 + $aData['sid']
 + $aData['oSurvey'] (to come)
 PS: We could include a this SID as a special attribute in the LSBAseController

 This survey id will be used for linking base buttons already included in some base top bar view and for permission checking.
 (see LayoutHelper::getSurveyTopbarData())

- Other parameters: topBar also has the following parameters which values are set by default in LayoutHelper::getSurveyTopbarData().
 These are overridable by setting them in the topBar array.
 + oSurvey
 + canactivate
 + expired
 + notstarted
 + context
 + contextbutton
 + language
 + sumcount
 + hasSurveyContentPermission
 + hasDeletePermission
 + hasSurveyTranslatePermission
 + hasAdditionalLanguages
 + conditionsCount
 + hasSurveyReadPermission
 + hasSurveyTokensPermission
 + hasResponsesCreatePermission
 + hasResponsesReadPermission
 + hasSurveyActivationPermission
 + hasResponsesStatisticsReadPermission
 + extraToolsMenuItems
 + beforeSurveyBarRender
 + showToolsMenu

 PS: Refactoring still needed inside LayoutHelper::getSurveyTopbarData()

**Comment by olleharstedt** [comment]
Travis is red?

**Comment by olleharstedt** [comment]
Candidates for classes:

* TopbarConfiguration (containing data (no behaviour))
* TopbarWidget (containing all the view and layout files)

**Comment by olleharstedt** [comment]
With a class, you can set default values in the constructor or directly where the property is declared.

```php
class MyConfigClass
{
  private $id = 0;  // Default values
  public function __constructor(array $config)
  {
    // Default values can also be set like this
    $this->id = isset($config['customId']) ? $config['customId'] : 0;
  }
}

**Comment by gabrieljenik** [comment]
> Travis is red?

This is because some topbar still don't have a save button.
```
1) ls\tests\QuestionGroupEditorTest::testCreateQuestionGroup
Facebook\WebDriver\Exception\NoSuchElementException: Unable to locate element: #save\-button
``` 

I would just leave it for now and pick that up later.

**Comment by gabrieljenik** [comment]
> Thanks for presenting your solution, Gabriel!
> 
> Candidates for classes:
> 
> * TopbarConfiguration (containing data, no behaviour)
> * TopbarWidget (containing behaviour, like checking permissions)

So you are saying to turn the rendering into a widget and the configuration into a class, right?

**Comment by olleharstedt** [comment]
> So you are saying to turn the rendering into a widget and the configuration into a class, right?

Yeah, the base layout in the widget, then specific layouts in controller view folders. But you have to test and see what makes sense. :) Hard to predict without actually working with the code.

Related (for configuration class): https://en.wikipedia.org/wiki/Value_object

**Comment by gabrieljenik** [comment]
Hi All ! 

I have just pushed some changes about turning the configuration stuff into a class.
I know I have said I would send the interface before starting, but while reviewing it and trying some stuff, we found outselves very close to the implementation.

A bit of changes on the configuration are still pending (like a `originalConfig` attribute).
We will tackle that when switching to the widget.

Look forward to your comments.

Thanks!

File: application/core/TopbarConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    /**
+     * This Method is returning the Data for Survey Top Bar
+     *
+     * @param int $sid Given Survey ID
+     *
+     * @return array
+     * @throws CException
+     *
+     */
+    protected function getSurveyTopbarData($sid) {
```
Comment (olleharstedt) [inline application/core/TopbarConfiguration.php:72]: PSR-12

File: application/helpers/LayoutHelper.php [hunk 2]
```diff
@@ -452,6 +452,27 @@ protected function beforeSideMenuRender(array $aData)
...
+     */
+    public static function renderTopbar($aData) {
+
+        $oTopbarConfig = TopbarConfiguration::fromViewData($aData);
+
+        if (!empty($oTopbarConfig->leftSideView)) $aData['topBar']['leftSideContent'] = App()->getController()->renderPartial($oTopbarConfig->leftSideView, $aData, true);
+        if (!empty($oTopbarConfig->rightSideView)) $aData['topBar']['rightSideContent'] = App()->getController()->renderPartial($oTopbarConfig->rightSideView, $aData, true);
+
+        // Temporary fix before implementing the TopbarWidget
+        $aData['topBar']['name'] = $oTopbarConfig->viewName;
```
Comment (olleharstedt) [inline application/helpers/LayoutHelper.php:466]: Why not just send the topbar config object to the views? `$aData['topBar']['topbarConfig'] = $oTopbarConfig;`

**Review by olleharstedt — CHANGES_REQUESTED** [review]

**Comment by olleharstedt** [comment]
Thanks for your changes, Gabriel! I added a couple of comments.

File: application/core/TopbarConfiguration.php [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    private $surveyData = array();
+
+    /** @var string Name of the view used to render the left side of the topbar */
+    private $leftSideView = '';
+
+    /** @var string Name of the view used to render the right side of the topbar */
+    private $rightSideView = '';
+
+    /** @var array List of properties that can have public getter */
+    private static $publicReadableProperties = ['viewName', 'id', 'data', 'leftSideView', 'rightSideView', 'surveyData'];
```
[the remark this answers is not in the document]
Comment (gabrieljenik) [reply application/core/TopbarConfiguration.php:24]: haha this seems to be more efficient !!

File: application/helpers/LayoutHelper.php (hunk 2 above)
Comment (gabrieljenik) [reply application/helpers/LayoutHelper.php:466]: We will when we implement the widget. Didn't want to go to far on this intermediate step.

Comment (olleharstedt) [reply application/helpers/LayoutHelper.php:466]: Alright

File: application/core/TopbarConfiguration.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (olleharstedt) [reply application/core/TopbarConfiguration.php:24]: It's more efficient to _write_, but not more efficient to _read_. The latter is usually more important.

**Comment by gabrieljenik** [comment]
Amended some of the comments by @olleharstedt 
Made the rendering to be in a widget

Next step if you are fine, review all screens and make sure they have their topbars

**Comment by cdorin93** [comment]
@olleharstedt & @Trischi80 : what is the status of this PR :thinking:

**Comment by olleharstedt** [comment]
1. Give Gabriel time to finish it to remove all Vue top bars
2. Make Travis green and happy
3. Merge!

**Comment by gabrieljenik** [comment]
Agree.

**Comment by gabrieljenik** [comment]
Need to review some stuff still, but spreaded the top bar among many controllers.

**Comment by gabrieljenik** [comment]
Added Topbar for Check Logic and Condition Designer

**Comment by gabrieljenik** [comment]
The question editor changes from overview to edit view without refreshing, how is that supposed to work with topbars?
What are your expectations?

**Comment by olleharstedt** [comment]
> The question editor changes from overview to edit view without refreshing, how is that supposed to work with topbars?
> What are your expectations? @olleharstedt

There's a simple javascript snippet run when switching, you could just expand on that.

**Comment by gabrieljenik** [comment]
@olleharstedt I think this is ready to go.
Please, can I ask you to double check if the latest changes in the question controller are OK (there was some back adn froth in the merge with develop)

**Comment by olleharstedt** [comment]
Hi, can you merge in the latest dev branch, so Travis gets green?

**Comment by gabrieljenik** [comment]
Done. Let's see how Travis takes this.

**Comment by gabrieljenik** [comment]
Still on red. Will review it.

**Comment by olleharstedt** [comment]
> 1) ls\tests\SaveDualScaleAnswerOptionsTest::testQuestionEditor
> Facebook\WebDriver\Exception\NoSuchElementException: Unable to locate element: #save\-button\-create\-question

Looks like the id changed of the save button?

**Comment by gabrieljenik** [comment]
> > 1. ls\tests\SaveDualScaleAnswerOptionsTest::testQuestionEditor
> >    Facebook\WebDriver\Exception\NoSuchElementException: Unable to locate element: #save-button-create-question
> 
> Looks like the id changed of the save button?

Could have as it is a little bit more standarized. Have to review.

**Comment by olleharstedt** [comment]
> Could have as it is a little bit more standarized. Have to review.

You "standardized" the element ids that all Javascript depends on? :laughing:

**Comment by gabrieljenik** [comment]
> > Could have as it is a little bit more standarized. Have to review.
> 
> You "standardized" the element ids that all Javascript depends on?  Don't make it harder than it have to be.

Not all. There is a defualt. Overridable. This could have not got overriden.

**Comment by gabrieljenik** [comment]
We got a green!!! Merge quickly before it turns red again!

**Comment by olleharstedt** [comment]
> Pushed WIP from the surveyTopbar.
> If no top bar is defined, surveyTopbar_view is used.
> image
> # Some details about topBar
> 
> topBar is an array with (so far) the following parameters:
> 
>     * name: Name of the view used for rendering the topbar.
>       Some already created views:
>       
>       * baseTopbar_view.
>         This is a blank topBar which allows left and right subviews.
>       * surveyTopbar_view.
>         This is the default topBar which already includes base buttons for the survey being used.
>         Base buttons will be linked to the survey
> 
>     * leftSideView and rightSideView: views used for rendering the left side and right side of the topbar.
> 
>     * sid: When using surveyTopbar_view, this is surveyId of the survey being used.
>       If not specifically set, the helper will look for the survey id from the following:
> 
> 
>     * $aData['sid']
> 
>     * $aData['oSurvey'] (to come)
>       PS: We could include a this SID as a special attribute in the LSBAseController
> 
> 
> This survey id will be used for linking base buttons already included in some base top bar view and for permission checking.
> (see LayoutHelper::getSurveyTopbarData())
> 
>     * Other parameters: topBar also has the following parameters which values are set by default in LayoutHelper::getSurveyTopbarData() -- this was ported from LSBaseController. Should be removed from there eventually (or refactored).
>       These are overridable by setting them in the topBar array.
> 
> 
>     * oSurvey
> 
>     * canactivate
> 
>     * expired
> 
>     * notstarted
> 
>     * context
> 
>     * contextbutton
> 
>     * language
> 
>     * sumcount
> 
>     * hasSurveyContentPermission
> 
>     * hasDeletePermission
> 
>     * hasSurveyTranslatePermission
> 
>     * hasAdditionalLanguages
> 
>     * conditionsCount
> 
>     * hasSurveyReadPermission
> 
>     * hasSurveyTokensPermission
> 
>     * hasResponsesCreatePermission
> 
>     * hasResponsesReadPermission
> 
>     * hasSurveyActivationPermission
> 
>     * hasResponsesStatisticsReadPermission
> 
>     * extraToolsMenuItems
> 
>     * beforeSurveyBarRender
> 
>     * showToolsMenu
> 
> 
> PS: Refactoring still needed inside LayoutHelper::getSurveyTopbarData()

Btw, mega-arrays like these are why DTO are preferable. We have them everywhere in the code base right now. :|

**Review by olleharstedt — APPROVED** [review]

**Pull request merged**